onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc

```diff
@@ -117,22 +117,45 @@ NnapiExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph_view
     }
   }
 
+  // Get all the NodeUnits in the graph_viewer
+  std::vector<std::unique_ptr<NodeUnit>> node_unit_holder;
+  std::unordered_map<const Node*, const NodeUnit*> node_unit_map;
+
+  std::tie(node_unit_holder, node_unit_map) = GetAllNodeUnits(graph_viewer);
+
+  // This holds the result of whether a NodeUnit is supported or not,
+  // to prevent nodes in a NodeUnit to be checked for multiple times
+  std::unordered_map<const NodeUnit*, bool> node_unit_supported_result;
+  node_unit_supported_result.reserve(node_unit_holder.size());
+
   const auto excluded_nodes = utils::CreateExcludedNodeSet(graph_viewer, partitioning_stop_ops_);
   const bool check_excluded_nodes = !excluded_nodes.empty();
 
   std::unordered_set<std::string> node_outputs_in_current_group{};
 
   const auto is_node_supported = [&](const Node& node) -> bool {
-    const bool excluded = check_excluded_nodes && Contains(excluded_nodes, &node);
-    // TODO, cache NodeUnit(s) and not generate new instance every time
-    const NodeUnit node_unit(node);
-    const bool supported = !excluded &&
-                           nnapi::IsNodeSupportedInGroup(node_unit, graph_viewer, params,
-                                                         node_outputs_in_current_group);
-    LOGS_DEFAULT(VERBOSE) << "Operator type: [" << node_unit.OpType()
-                          << "] index: [" << node_unit.Index()
-                          << "] name: [" << node_unit.Name()
-                          << "] supported: [" << supported
+    const NodeUnit* node_unit = node_unit_map.at(&node);
+    bool supported = false;
+
+    // If we have visited one of the nodes in the node_unit, use the result directly
+    if (Contains(node_unit_supported_result, node_unit)) {
+      supported = node_unit_supported_result.at(node_unit);
```

Comment:
>at [](http://example.com/codeflow?start=45&length=2)

the two lookups from Contains() and at() could be replaced with a single one from find()